casper/src/main/scala/coop/rchain/casper/finality/Finalizer.scala

```diff
@@ -0,0 +1,192 @@
+import coop.rchain.models.syntax._
+import coop.rchain.shared.Log
+import coop.rchain.shared.syntax._
+import fs2.Stream
+
+/**
+  * Block can be recorded as last finalized block (LFB) if Safety oracle outputs fault tolerance (FT)
+  * for this block greater then some predefined threshold. This is defined by [[CliqueOracle.computeOutput]]
+  * function, which requires some target block as input arg.
+  *
+  * Therefore: Finalizer has a scope of search, defined by tips and previous LFB - each of this blocks can be next LFB.
+  *
+  * We know that LFB advancement is not necessary continuous, next LFB might not be direct child of current one.
+  *
+  * Therefore: we cannot start from current LFB children and traverse DAG from the bottom to the top, calculating FT
+  * for each block. Also its computationally ineffective.
+  *
+  * But we know that scope of search for potential next LFB is constrained. Block A can be finalized only
+  * if it has more then half of total stake in bonds map of A translated from tips throughout main parent chain.
+  * IMPORTANT: only main parent relation gives weight to potentially finalized block.
+  *
+  * Therefore: Finalizer should seek for next LFB going through 2 steps:
+  *   1. Find messages in scope of search that have more then half of the stake translated through main parent chain
+  *     from tips down to the message.
+  *   2. Execute [[CliqueOracle.computeOutput]] on these targets.
+  *   3. First message passing FT threshold becomes the next LFB.
+  */
+object Finalizer {
+
+  private type WeightMap = Map[Validator, Long]
+
+  /** Message that is agreed on + weight of this agreement. */
+  final case class MessageAgreement(
+      message: BlockMetadata,
+      messageWeightMap: WeightMap,
+      stakeAgreed: WeightMap
+  )
+
+  /** weight map as per message, look inside [[CliqueOracle.getCorrespondingWeightMap]] description for more info */
+  private def messageWeightMapF[F[_]: Sync](
+      message: BlockMetadata,
+      dag: BlockDagRepresentation[F]
+  ): F[WeightMap] = CliqueOracle.getCorrespondingWeightMap(message.blockHash, dag)
+
+  /** If more then half of total stake agree on message - it is considered to be safe from orphaning. */
+  def cannotBeOrphaned[F[_]: Applicative](
+      messageWeightMap: WeightMap,
+      agreeingWeightMap: WeightMap
+  ): Boolean = {
+    require(
+      !agreeingWeightMap.exists { case (_, stake) => stake <= 0 },
+      message = "Agreeing map contains not bonded validators"
+    )
+    val activeStakeTotal    = messageWeightMap.values.sum
+    val activeStakeAgreeing = agreeingWeightMap.values.sum
+    // in theory if each stake is high enough, e.g. Long.MaxValue, sum of them might result in negative value
+    require(activeStakeTotal > 0, message = "Long overflow when computing total stake")
+    require(activeStakeAgreeing > 0, message = "Long overflow when computing total stake")
+    activeStakeAgreeing > activeStakeTotal.toFloat / 2
```

Comment:
How is this calculation here in finalizer? I don't see that was in previous calculation. This should be information from safety oracle.

Reply:
This is because previous finalizer were just running clique oracle on all LFB children. New version traverses down from tips until it finds messages worth to be computed faultTolerance agains. If you see clique oracle code - if less than half of a stake is agreed on message, max clique is not computed for this message, and minimal fault tolerance is returned straight away. So this message for sure cannot be finalized.

Do you mean this function better to be moved inside CliqueOracle?

Reply:
My concern is if we are sure here that agreeing validators cannot have equivocations.

Reply:
I'm not sure I understand this. There is always only one main parent, right? So agreeing chains are well defined. We need to get to the point where enough of these chains join, and find first message that pass fault toleranceThreshold.